db/kvstorewithbuffer.go

```diff
@@ -168,6 +168,24 @@ func (kvb *kvStoreWithBuffer) MustDelete(ns string, key []byte) {
 	kvb.buffer.Delete(ns, key, "failed to delete %x in %s", key, ns)
 }
 
+func (kvb *kvStoreWithBuffer) Filter(ns string, cond Condition) ([][]byte, [][]byte, error) {
+	fk, fv, err := kvb.store.Filter(ns, cond)
+	if err != nil {
+		return fk, fv, err
+	}
+
+	// filter the entries in buffer
+	for i := 0; i < kvb.buffer.Size(); i++ {
+		entry, _ := kvb.buffer.Entry(i)
+		k, v := entry.Key(), entry.Value()
```

Comment:
the entries in a buffer are not naive key-value pair. For "Delete", we need to remove it from fk and fv, and only do append if it is a "Put" type.